letsencrypt_auto/letsencrypt-auto.template

```diff
@@ -141,65 +81,146 @@ then
-if [ "$VERBOSE" = 1 ]  ; then
-  echo
-  $VENV_BIN/pip install -U setuptools
-  $VENV_BIN/pip install -U pip
-  $VENV_BIN/pip install -r "$LEA_PATH"/py26reqs.txt -U letsencrypt letsencrypt-apache
-  # nginx is buggy / disabled for now, but upgrade it if the user has
-  # installed it manually
-  if $VENV_BIN/pip freeze | grep -q letsencrypt-nginx ; then
-    $VENV_BIN/pip install -U letsencrypt letsencrypt-nginx
+  if [ ! -f $VENV_BIN/letsencrypt ]; then
+    OLD_VERSION="0.0.0"  # ($VENV_BIN/letsencrypt --version)
+  else
+    OLD_VERSION="0.0.0"
   fi
-else
-  $VENV_BIN/pip install -U setuptools > /dev/null
-  printf .
-  $VENV_BIN/pip install -U pip > /dev/null
-  printf .
-  # nginx is buggy / disabled for now...
-  $VENV_BIN/pip install -r "$LEA_PATH"/py26reqs.txt > /dev/null
-  printf .
-  $VENV_BIN/pip install -U letsencrypt > /dev/null
-  printf .
-  $VENV_BIN/pip install -U letsencrypt-apache > /dev/null
-  if $VENV_BIN/pip freeze | grep -q letsencrypt-nginx ; then
-    printf .
-    $VENV_BIN/pip install -U letsencrypt-nginx > /dev/null
+
+  # TODO: Don't bother upgrading if we're already up to date.
+  if [ "$OLD_VERSION" != "1.2.3" ]; then
+    Bootstrap
+    echo "Creating virtual environment..."
+    rm -rf "$VENV_PATH"
+    DeterminePythonVersion
+    if [ "$VERBOSE" = 1 ] ; then
+      virtualenv --no-site-packages --python $LE_PYTHON $VENV_PATH
+    else
+      virtualenv --no-site-packages --python $LE_PYTHON $VENV_PATH > /dev/null
+    fi
+
+    # Now we drop into Python so we don't have to install even more
+    # dependencies (curl, etc.), for better flow control, and for the
+    # option of future Windows compatibility.
+    #
+    # This Python script prints a path to a temp dir
+    # containing a new copy of letsencrypt-auto or returns non-zero.
+    # There is no $ interpolation due to quotes on heredoc delimiters.
+    set +e
+    TEMP_DIR=`$LE_PYTHON - << "UNLIKELY_EOF"
+{{ download_upgrade.py }}
+UNLIKELY_EOF`
+    DOWNLOAD_STATUS=$?
+    set -e
+    if [ "$DOWNLOAD_STATUS" = 0 ]; then
+      # Install new copy of letsencrypt-auto. This preserves permissions and
+      # ownership from the old copy.
+      # TODO: Deal with quotes in pathnames.
+      echo "Installing new version of letsencrypt-auto..."
+      echo "  " $SUDO cp "$TEMP_DIR/letsencrypt-auto" "$0"
+      $SUDO cp "$TEMP_DIR/letsencrypt-auto" "$0"
+      # TODO: Clean up temp dir safely, even if it has quotes in its path.
+      "$0" --_skip-to-install "$TEMP_DIR" "$@"
+    else
+      # Report error:
+      echo $TEMP_DIR
+      exit 1
+    fi
+  fi  # should upgrade
+else  # --_skip-to-install was passed.
+  # Install Python dependencies with peep, then run letsencrypt.
+  echo "Installing Python package dependencies..."
+  TEMP_DIR="$2"
+  shift 2
+  cat << "UNLIKELY_EOF" > $TEMP_DIR/letsencrypt-auto-requirements.txt
+{{ letsencrypt-auto-requirements.txt }}
+UNLIKELY_EOF
+
+  cat << "UNLIKELY_EOF" > $TEMP_DIR/peep.py
```

Comment:
Perhaps add a big visual separator line here, because we're about to change languages and universes :)